### PR TITLE
test(state): mocking State interface

### LINKED ISSUE #2166: State

### PR DESCRIPTION
**Pull request by b00f**
## Description

This PR mocks the `State` interface using gomock.

## Related Issue

Fixes #2166

### CODE REVIEW

**Comment by b00f** [comment]
## Code Review: PR #2193

### Overall Assessment
This is a substantial and well-structured PR that replaces the hand-rolled `state/mock.go` with gomock-generated mocks and introduces `FakeState` as a test fixture. The overall direction is good — renaming `Facade` → `State`, extracting `NewManager` for testability, and the fake pattern are all solid improvements. However, there is a **critical bug** that causes `consensusv2` tests to fail, and several design issues worth addressing before merge.

---

## Critical

### 1. `state_fake.go`: `ProposeBlock` generates blocks at wrong height

**Location:** `state/state_fake.go:117-124`

```go
mock.EXPECT().ProposeBlock(gomock.Any(), gomock.Any()).DoAndReturn(
    func(valKey *bls.ValidatorKey, _ crypto.Address) (*block.Block, error) {
        blk, _ := ts.GenerateTestBlock(fake.LastHeight,  // ← BUG: should be LastHeight + 1
            testsuite.BlockWithProposer(valKey.Address()))
        return blk, nil
    },
).AnyTimes()
```

`GenerateTestBlock(fake.LastHeight, ...)` generates a block with `Height = LastHeight`. When consensus proposes the *next* block, the height should be `LastHeight + 1`. However, `CommitBlock` (line 128) checks `blk.Height() == fake.LastHeight+1`, which fails because the block is one height too low.

**Impact:** `consensusv2` tests fail with `"committing block failed error=invalid"` whenever consensus reaches the commit stage.

**Evidence:**
```
$ go test ./consensusv2/ -count=1 -run TestConsensusNormal
ERR committing block failed error=invalid
--- FAIL: TestConsensusNormal (0.14s)
    Error: Received unexpected error: unable to reach consensus
```

**Fix:** Change `fake.LastHeight` to `fake.LastHeight + 1` on line 119.

---

## Major

### 2. `FakeState` missing default mock setups for several `State` methods

The following methods have no `DoAndReturn` default: `IsProposer`, `PendingTx`, `AddPendingTx`, `AddPendingTxAndBroadcast`, `UpdateValidatorProtocolVersion`, `CalculateFee`, `PublicKey`, `AllPendingTxs`, `CheckTransaction`, `Close`.

**Risk:** `IsProposer` is called in consensus and may trigger "unexpected call" gomock panics when used without explicit `EXPECT()` setup. `CalculateFee` and `CheckTransaction` may also cause issues in test flows that need default values.

**Recommendation:** At minimum, add `DoAndReturn` for `IsProposer` that delegates to the committee. Consider sensible defaults for `CalculateFee` (return a small fixed value) and `CheckTransaction` (return nil).

---

## Minor

### 3. `FakeGRPCServer`: confusing duplicate `FakeState`/`MockState` fields

**Location:** `www/grpc/fake/server_fake.go:19-20`

Both fields `FakeState *state.FakeState` and `MockState *state.FakeState` point to the same object. Tests use them interchangeably — see `blockchain_test.go:324` (`FakeState.EXPECT()`) vs `:397` (`MockState.EXPECT()`). This is confusing.

**Recommendation:** Drop one field. Since `FakeState` embeds `*MockState`, a single field suffices. Rename all usages consistently.

### 4. Thin `Fake*` wrappers add no behavioral simulation

`consensus/consensus_fake.go`, `consensus/manager/manager_fake.go`, `wallet/manager/manager_fake.go`, and `sync/sync_fake.go` are ~15-line wrappers that embed gomock mocks with zero default behavior. They add an unnecessary abstraction layer.

**Recommendation:** Either add meaningful defaults (like `FakeState` does) or drop these wrappers and use `NewMock*()` directly. The "Fake" name should match the substance.

### 5. Inconsistent constructor patterns

`sync/sync_fake.go` uses `MockingSync(ts)` while all other `NewFake*` use `NewMock*(ts.MockController())`. Different patterns for the same purpose.

**Recommendation:** Standardize — either move `MockingSync` setup into `NewFakeSync`, or refactor `MockingSync` to follow the `NewMockSync` + configuration pattern.

### 6. Hardcoded magic number in `ProposerIndex`

**Location:** `state/state_fake.go:298-302`
```go
func (f *FakeState) ProposerIndex(round types.Round) int {
    i := int(f.LastHeight%4) + int(round%4)
    return i % 4
}
```
The `% 4` assumes committee size of 4. This silently produces wrong results for tests with different committee sizes.

**Fix:** Use `len(f.Committee.Validators())`.

---

## Positive observations

- `Facade` → `State` rename is a clear improvement
- `IManager` → `WalletManager` rename is also good
- `NewManager` extraction makes the consensus manager properly testable
- `FakeState.AddTestBlock/AddTestAccount/AddTestValidator` helpers are excellent
- Sync handler test improvements (new edge case tests, `SafeDecrease`) are good
- The gomock integration via `testsuite.MockController()` is clean

---

The **critical bug must be fixed** before this can pass CI. The major and minor items are recommendations.

File: state/state_fake.go [hunk 1]
```diff
@@ -0,0 +1,334 @@
...
+
+	mock.EXPECT().AvailabilityScore(gomock.Any()).DoAndReturn(
+		func(int32) float64 {
+			return fake.Score
+		},
+	).AnyTimes()
+
+	mock.EXPECT().ProposeBlock(gomock.Any(), gomock.Any()).DoAndReturn(
+		func(valKey *bls.ValidatorKey, _ crypto.Address) (*block.Block, error) {
+			blk, _ := ts.GenerateTestBlock(fake.LastHeight,
```
Comment (b00f) [inline state/state_fake.go]: Critical Bug: GenerateTestBlock should use fake.LastHeight + 1, not fake.LastHeight.

The consensus calls ProposeBlock to create the next block. The real State.ProposeBlock generates a block at LastBlockHeight() + 1. But here, the block is generated at fake.LastHeight (current height), which is one less than expected.

CommitBlock at line 128 checks blk.Height() == fake.LastHeight + 1, and this check fails because the block's height is fake.LastHeight, not fake.LastHeight + 1.

This causes all consensusv2 tests to fail. Eg: 'go test ./consensusv2/ -run TestConsensusNormal' fails with 'unable to reach consensus'.

Fix: Change fake.LastHeight to fake.LastHeight + 1.

File: state/state_fake.go [hunk 2]
```diff
@@ -0,0 +1,334 @@
...
+				AverageScore:    fake.Score,
+			}
+		},
+	).AnyTimes()
+
+	return fake
+}
+
+func (f *FakeState) ProposerIndex(round types.Round) int {
+	i := int(f.LastHeight%4) + int(round%4)
```
Comment (b00f) [inline state/state_fake.go]: Minor: Hardcoded committee size of 4. This silently produces incorrect results for tests using different committee sizes.

Fix: Use `len(f.Committee.Validators())` instead.

Comment (b00f) [inline state/state_fake.go]: Minor: Hardcoded committee size of 4. This silently produces incorrect results for tests using different committee sizes.

Fix: Use `len(f.Committee.Validators())` instead.

File: www/grpc/fake/server_fake.go [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+	"github.com/pactus-project/pactus/sync"
+	"github.com/pactus-project/pactus/util"
+	"github.com/pactus-project/pactus/util/testsuite"
+	wltmgr "github.com/pactus-project/pactus/wallet/manager"
+	"github.com/pactus-project/pactus/www/grpc"
+	"github.com/pactus-project/pactus/www/zmq"
+)
+
+type FakeGRPCServer struct {
+	FakeState     *state.FakeState
```
Comment (b00f) [inline www/grpc/fake/server_fake.go:19]: Minor: Both `FakeState` and `MockState` point to the same `*state.FakeState` object. The tests use them interchangeably (see `blockchain_test.go` line 186 uses `MockState` for `AddTestAccount`, line 324 uses `FakeState` for `EXPECT()`). This is confusing and inconsistent.

Recommendation: Since `FakeState` embeds `*MockState` (so it supports both test helpers and EXPECT), pick one field name and use it consistently. Drop the other.

**Comment by b00f** [comment]
## Re-review: all critical issues fixed ✅

Verified the following fixes in the new commits:

### Fixed ✅

1. **`state_fake.go:125` — `ProposeBlock` height**: Changed `fake.LastHeight` → `fake.LastHeight + 1`. ConsensusV2 tests now pass.

2. **`state_fake.go:305-309` — `ProposerIndex` hardcoded `% 4`**: Changed to `len(f.Committee.Validators())`. Works for any committee size.

3. **`www/grpc/fake/server_fake.go` — Confusing `MockState`/`Mock*` fields**: Removed the duplicate `MockState` field, renamed all `MockSync`/`MockCons`/`MockConsMgr`/`MockWalletMgr` → `FakeSync`/`FakeCons`/`FakeConsMgr`/`FakeWalletMgr` for consistency. All test references updated accordingly.

### Test results

```
ok  github.com/pactus-project/pactus/state          (cached)
ok  github.com/pactus-project/pactus/consensus       (cached)
ok  github.com/pactus-project/pactus/consensus/manager  (cached)
ok  github.com/pactus-project/pactus/consensusv2     (cached)
ok  github.com/pactus-project/pactus/sync            (cached)
ok  github.com/pactus-project/pactus/www/grpc        (cached)
ok  github.com/pactus-project/pactus/www/html        (cached)
```

All tests pass. The PR is good to merge from a correctness standpoint.

**Review by themantre — APPROVED** [review]
